pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -17,20 +17,59 @@ limitations under the License.
 package persistentvolume
 
 import (
+	"fmt"
 	"time"
 
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/client/cache"
 	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 	"k8s.io/kubernetes/pkg/cloudprovider"
 	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/conversion"
 	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/volume"
 	"k8s.io/kubernetes/pkg/watch"
 
 	"github.com/golang/glog"
 )
 
+// Design:
+//
+// The fundamental key to this design is the bi-directional "pointer" between
+// PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs), which is
+// represented here as pvc.Spec.VolumeName and pv.Spec.ClaimRef. The bi-directionality
+// is complicated to manage in a transactionless system, but without it we
+// can't ensure sane behavior in the face of different forms of trouble.  For
+// example, a rogue HA controller instance could end up racing and making
+// multiple bindings that are indistinguishable, resulting in potential data
+// loss.
+//
+// This controller supports pre-bound (by the creator) objects in both
+// directions: a PVC that wants a specific PV or a PV that is reserved for a
+// specific PVC.
+//
+// The binding is two-step process. PV.Spec.ClaimRef is modified first and
+// PVC.Spec.VolumeName second. At any point of this transaction, the PV or PVC
+// can be modified by user or other controller or completelly deleted. Also, two
+// (or more) controllers may try to bind different volumes to different claims
+// at the same time. The controller must recover from any conflicts that may
+// arise from these conditions.
```

Comment:
It does not matter on the order of these objects. If claim "wants" a specific PV, it should bind only to the PV. If PV "wants" a specific claim, it will bind only to the claim. This means that your pv1 can bind only to claim1, which is not possible as claim1 can bind only to pv2. I.e. pv1 will be Pending forever. However, it is possible to bind claim1 to pv2, assuming that pv2 allows it. If pv2.Spec.ClaimRef.Name="claim2", claim1 will be pending forever too.